src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs

```diff
@@ -4972,35 +4969,122 @@ .maxstack  6
 ");
         }
 
-        [WorkItem(732269, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/732269")]
         [Fact]
+        [WorkItem(732269, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/732269")]
+        [WorkItem("https://github.com/dotnet/roslyn/issues/56007")]
         public void NullCoalesce()
         {
-            var source = @"
-class Program
-{
-    static int Main(int? x, int y) { return x ?? y; }
-}
-";
+            var source = """
+                class Program
+                {
+                    static int Main(int? x, int y) { return x ?? y; }
+                }
+                """;
+
             var comp = CompileAndVerify(source);
             comp.VerifyDiagnostics();
-            comp.VerifyIL("Program.Main", @"
-{
-  // Code size       21 (0x15)
-  .maxstack  1
-  .locals init (int? V_0)
-  IL_0000:  ldarg.0
-  IL_0001:  stloc.0
-  IL_0002:  ldloca.s   V_0
-  IL_0004:  call       ""bool int?.HasValue.get""
-  IL_0009:  brtrue.s   IL_000d
-  IL_000b:  ldarg.1
-  IL_000c:  ret
-  IL_000d:  ldloca.s   V_0
-  IL_000f:  call       ""int int?.GetValueOrDefault()""
-  IL_0014:  ret
-}
-");
+            comp.VerifyIL("Program.Main", """
+                {
+                  // Code size        9 (0x9)
+                  .maxstack  2
+                  IL_0000:  ldarga.s   V_0
+                  IL_0002:  ldarg.1
+                  IL_0003:  call       "int int?.GetValueOrDefault(int)"
+                  IL_0008:  ret
+                }
+                """);
+        }
+
+        [Theory, WorkItem("https://github.com/dotnet/roslyn/issues/56007")]
+        [InlineData("ref")]
+        [InlineData("in")]
+        [InlineData("ref readonly")]
+        public void NullCoalesce_RefParameter(string refKeyword)
+        {
+            var source = $$"""
+                class Program
+                {
+                    static int Main(int? x, {{refKeyword}} int y) { return x ?? y; }
+                }
+                """;
+
+            var comp = CompileAndVerify(source);
+            comp.VerifyDiagnostics();
+            comp.VerifyIL("Program.Main", """
+                {
+                  // Code size       10 (0xa)
+                  .maxstack  2
+                  IL_0000:  ldarga.s   V_0
+                  IL_0002:  ldarg.1
+                  IL_0003:  ldind.i4
```

Comment:
An attempt to dereference might cause an exception. We probably shouldn't optimize scenarios like this.

Reply:
Ok, will exclude all non-ref locals/parameters from this optimization. But out of curiosity, in what cases dereferencing throws?

Reply:
>  in what cases dereferencing throws?

An access violation, for example.

Reply:
Did you mean "exclude all ref locals/parameters"?

Reply:
> Did you mean "exclude all ref locals/parameters"?

I had 2 thoughts at that moment: "include only non-ref locals/parameters" and "exclude all ref locals/parameters". These 2 thoughts collided and we got a merge conflict)